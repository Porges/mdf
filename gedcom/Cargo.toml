[package]
name = "gedcom"
version = "0.1.0"
edition = "2021"
authors = ["George Pollard <porges@porg.es>"]
license = "GPL-3.0"

[features]
default = ["miette-highlighting"]
miette-highlighting = ["miette/fancy-no-backtrace"]
legacy-encodings = ["dep:oem_cp"]

[dependencies]
ascii = "1.1.0"
encoding_rs = "0.8.34"
# TODO: update once this is merged https://github.com/zkat/miette/pull/393
miette = { git = "https://github.com/Porges/miette", branch = "invalid_unicode" }
oem_cp = { version = "2.0.0", optional = true }
owo-colors = "4.0.0"
thiserror = "1.0.63"
tracing = { version = "0.1", features = ["attributes"] }
vec1 = "1.12.1"

[dev-dependencies]
insta = { version = "1.39.0", features = ["glob"] }
kdl = "4.6.0"
