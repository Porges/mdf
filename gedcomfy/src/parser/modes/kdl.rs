use kdl::{KdlDocument, KdlEntry, KdlNode};

use crate::parser::{
    lines::LineValue, records::RawRecord, AnySourceCode, NonFatalHandler, ParseMode, ResultBuilder,
    Sourced,
};

#[derive(Default)]
pub(in crate::parser) struct Mode {}

impl NonFatalHandler for Mode {
    fn non_fatal<E>(&mut self, _error: E) -> Result<(), E>
    where
        E: Into<crate::parser::ParseError> + miette::Diagnostic,
    {
        Ok(())
    }
}

impl ParseMode for Mode {
    type ResultBuilder<'i> = Builder;

    fn get_result_builder<'i>(
        self,
        _version: crate::versions::SupportedGEDCOMVersion,
        _source_code: AnySourceCode,
    ) -> Result<Self::ResultBuilder<'i>, crate::parser::ParseError> {
        Ok(Builder {
            mode: self,
            doc: KdlDocument::new(),
        })
    }
}

pub(in crate::parser) struct Builder {
    mode: Mode,
    doc: KdlDocument,
}

impl NonFatalHandler for Builder {
    fn non_fatal<E>(&mut self, error: E) -> Result<(), E>
    where
        E: Into<crate::parser::ParseError> + miette::Diagnostic,
    {
        self.mode.non_fatal(error)
    }
}

impl<'i> ResultBuilder<'i> for Builder {
    type Result = KdlDocument;

    fn handle_record(
        &mut self,
        record: Sourced<RawRecord<'i>>,
    ) -> Result<(), crate::parser::ParseError> {
        self.doc.nodes_mut().push(record_to_kdl(record.value));
        Ok(())
    }

    fn complete(self) -> Result<Self::Result, crate::parser::ParseError> {
        Ok(self.doc)
    }
}

fn record_to_kdl(record: RawRecord) -> KdlNode {
    let mut node = KdlNode::new(record.line.tag.to_string());

    if let Some(xref) = &record.line.xref {
        node.entries_mut()
            .push(KdlEntry::new_prop("xref", xref.value.to_string()));
    }

    if let Some(mapped) = match record.line.line_value.value {
        LineValue::Ptr(None) => Some(KdlEntry::new_prop("see", kdl::KdlValue::Null)),
        LineValue::Ptr(Some(value)) => Some(KdlEntry::new_prop("see", value)),
        LineValue::Str(data) => Some(KdlEntry::new(data.to_string())),
        LineValue::None => None,
    } {
        node.entries_mut().push(mapped);
    }

    if record.records.is_empty() {
        return node;
    }

    let mut children = KdlDocument::new();
    for subrecord in record.records {
        children.nodes_mut().push(record_to_kdl(subrecord.value));
    }

    node.set_children(children);
    node
}
