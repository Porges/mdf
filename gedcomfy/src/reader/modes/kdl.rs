use kdl::{KdlDocument, KdlEntry, KdlNode};

use crate::reader::{
    lines::LineValue, records::RawRecord, NonFatalHandler, ReadMode, ResultBuilder, Sourced,
};

#[derive(Default)]
pub(in crate::reader) struct Mode {}

impl NonFatalHandler for Mode {
    fn report<E>(&mut self, _error: E) -> Result<(), E>
    where
        E: Into<crate::reader::ReaderError> + miette::Diagnostic,
    {
        Ok(())
    }
}

impl<'i> ReadMode<'i> for Mode {
    type ResultBuilder = Builder;

    fn into_result_builder(
        self,
        _version: crate::versions::SupportedGEDCOMVersion,
    ) -> Result<Self::ResultBuilder, crate::reader::ReaderError> {
        Ok(Builder {
            mode: self,
            doc: KdlDocument::new(),
        })
    }
}

pub(in crate::reader) struct Builder {
    mode: Mode,
    doc: KdlDocument,
}

impl NonFatalHandler for Builder {
    fn report<E>(&mut self, error: E) -> Result<(), E>
    where
        E: Into<crate::reader::ReaderError> + miette::Diagnostic,
    {
        self.mode.report(error)
    }
}

impl<'i> ResultBuilder<'i> for Builder {
    type Result = KdlDocument;

    fn handle_record(
        &mut self,
        record: Sourced<RawRecord>,
    ) -> Result<(), crate::reader::ReaderError> {
        self.doc
            .nodes_mut()
            .push(record_to_kdl(record.sourced_value));
        Ok(())
    }

    fn complete(self) -> Result<Self::Result, crate::reader::ReaderError> {
        Ok(self.doc)
    }
}

fn record_to_kdl(record: RawRecord) -> KdlNode {
    let mut node = KdlNode::new(record.line.tag.to_string());

    if let Some(xref) = &record.line.xref {
        node.entries_mut()
            .push(KdlEntry::new_prop("xref", xref.sourced_value.to_string()));
    }

    if let Some(mapped) = match record.line.value.sourced_value {
        LineValue::Ptr(None) => Some(KdlEntry::new_prop("see", kdl::KdlValue::Null)),
        LineValue::Ptr(Some(value)) => Some(KdlEntry::new_prop("see", value)),
        LineValue::Str(data) => Some(KdlEntry::new(data.to_string())),
        LineValue::None => None,
    } {
        node.entries_mut().push(mapped);
    }

    if record.records.is_empty() {
        return node;
    }

    let mut children = KdlDocument::new();
    for subrecord in record.records {
        children
            .nodes_mut()
            .push(record_to_kdl(subrecord.sourced_value));
    }

    node.set_children(children);
    node
}
