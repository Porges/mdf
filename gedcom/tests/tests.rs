use std::{path::PathBuf, sync::Once};

use gedcom::{
    parser::{
        decoding,
        encodings::SupportedEncoding,
        options::{OptionSetting, ParseOptions},
        parse,
        records::RawRecord,
    },
    versions::GEDCOMVersion,
};
use kdl::{KdlDocument, KdlEntry, KdlNode};
use miette::{NamedSource, Report};

static INIT: Once = Once::new();
fn ensure_hook() {
    INIT.call_once(|| {
        miette::set_hook(Box::new(|_diag| {
            Box::new(
                miette::MietteHandlerOpts::new()
                    .terminal_links(false)
                    .unicode(true)
                    .color(false)
                    .width(132)
                    .build(),
            )
        }))
        .unwrap();
    });
}

#[test]
fn can_parse_allged_lines() {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("tests/external/gpdf/allged.ged");

    let data = std::fs::read(path).unwrap();

    let line_count = gedcom::validate_syntax(&data).unwrap();
    assert_eq!(line_count, 1159);
}

#[test]
fn produces_expected_allged_tree() {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("tests/external/gpdf/allged.ged");

    let data = std::fs::read(path).unwrap();
    let buffer = &mut String::new();
    let parsed = parse(&data, buffer, LENIENT_OPTIONS).unwrap();

    insta::assert_snapshot!(to_kdl(parsed.into_iter().map(|r| r.value)));
}

const LENIENT_OPTIONS: &ParseOptions = &ParseOptions {
    version: OptionSetting::Assume(GEDCOMVersion::V5),
    encoding: OptionSetting::Assume(SupportedEncoding::ANSEL),
};

#[test]
fn torture_test_valid() {
    ensure_hook();

    insta::glob!("external/gedcom-samples/Torture Test/*.ged", |path| {
        let data = std::fs::read(path).unwrap();
        let filename = path.file_name().unwrap().to_string_lossy();
        let buffer = &mut String::new();
        let parsed = parse(&data, buffer, LENIENT_OPTIONS)
            .map_err(|e| Report::new(e).with_source_code(NamedSource::new(filename, data.clone())))
            .unwrap();
        insta::assert_snapshot!(to_kdl(parsed.into_iter().map(|r| r.value)));
    });
}

#[test]
fn golden_files() -> miette::Result<()> {
    ensure_hook();

    insta::glob!("syntax_inputs/*.ged", |path| {
        let data = std::fs::read(path).unwrap();
        let filename = path.file_name().unwrap();
        insta::with_settings!({
            // provide GEDCOM source alongside output
            description => String::from_utf8_lossy(&data),
        }, {
            match parse(&data, &mut String::new(), LENIENT_OPTIONS) {
                Ok(records) => {
                    let kdl = to_kdl(records.into_iter().map(|r| r.value));
                    insta::assert_snapshot!(kdl);
                }
                Err(err) => insta::assert_snapshot!(format!(
                    "{:?}",
                    Report::new(err).with_source_code(NamedSource::new(filename.to_string_lossy(), data))
                )),
            };
        });
    });

    insta::glob!("format_inputs/*.ged", |path| {
        let data = std::fs::read(path).unwrap();
        let filename = path.file_name().unwrap().to_string_lossy();
        insta::with_settings!({
            // provide GEDCOM source alongside output
            description => String::from_utf8_lossy(&data),
        }, {
            match parse(&data, &mut String::new(), LENIENT_OPTIONS) {
                Ok(records) => {
                    let kdl = to_kdl(records.into_iter().map(|r| r.value));
                    insta::assert_snapshot!(kdl);
                }
                Err(err) => insta::assert_snapshot!(format!(
                    "{:?}",
                    Report::new(err).with_source_code(NamedSource::new(filename, data))
                )),
            };
        });
    });

    Ok(())
}

fn to_kdl<'a>(records: impl Iterator<Item = RawRecord<'a>>) -> KdlDocument {
    let mut doc = KdlDocument::new();
    for record in records {
        doc.nodes_mut().push(record_to_kdl(record));
    }

    doc
}

fn record_to_kdl(record: RawRecord) -> KdlNode {
    let mut node = KdlNode::new(record.line.tag.to_string());

    if let Some(xref) = &record.line.xref {
        node.entries_mut()
            .push(KdlEntry::new_prop("xref", xref.value.to_string()));
    }

    if let Some(data) = &record.line.data {
        node.entries_mut()
            .push(KdlEntry::new(data.value.to_string()));
    }

    if record.records.is_empty() {
        return node;
    }

    let mut children = KdlDocument::new();
    for subrecord in record.records {
        children.nodes_mut().push(record_to_kdl(subrecord.value));
    }

    node.set_children(children);
    node
}

#[test]
fn test_encodings() {
    ensure_hook();

    insta::glob!("encoding_inputs/*.ged", |path| {
        let data = std::fs::read(path).unwrap();
        let data_ref: &[u8] = data.as_ref();
        let filename = path.file_name().unwrap().to_string_lossy();
        let encoding_report =
            match decoding::version_and_encoding_from_gedcom(data_ref, LENIENT_OPTIONS) {
                Ok((_version, detected)) => format!(
                    "Encoding detected: {}\nReason: {}",
                    detected.encoding,
                    Report::new(detected.reason)
                        .with_source_code(NamedSource::new(filename.clone(), data.clone()))
                ),
                Err(err) => format!(
                    "{}",
                    Report::new(err)
                        .with_source_code(NamedSource::new(filename.clone(), data.clone()))
                ),
            };

        insta::with_settings!({
            // provide GEDCOM source alongside output
            description => String::from_utf8_lossy(&data),
        }, {
            insta::assert_snapshot!(encoding_report);
            match parse(&data, &mut String::new(), &LENIENT_OPTIONS) {
                Ok(records) => {
                    let kdl = to_kdl(records.into_iter().map(|r| r.value));
                    insta::assert_snapshot!(kdl);
                }
                Err(err) => insta::assert_snapshot!(format!(
                    "{:?}",
                    Report::new(err).with_source_code(NamedSource::new(filename, data))
                )),
            };
        });
    });
}
