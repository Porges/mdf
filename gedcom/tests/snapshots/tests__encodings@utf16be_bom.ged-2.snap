---
source: gedcom/tests/tests.rs
description: "��\u00000\u0000 \u0000H\u0000E\u0000A\u0000D\u0000\r\u0000\n\u00000\u0000 \u0000T\u0000R\u0000L\u0000R\u0000\r\u0000\n"
expression: kdl
input_file: gedcom/tests/encoding_inputs/utf16be_bom.ged
---
HEAD
TRLR
